cryptarchia/cryptarchia.py

```diff
@@ -118,37 +122,103 @@ def on_block(self, block: BlockHeader):
             self.forks.append(new_chain)
         # otherwise, we're missing the parent block
         # in that case, just ignore the block
-        
 
     # Evaluate the fork choice rule and return the block header of the block that should be the head of the chain
     def fork_choice(local_chain: Chain, forks: List[Chain]) -> BlockHeader:
         pass
-        
 
     def tip(self) -> BlockHeader:
         return self.fork_choice()
 
 
 @dataclass
 class Coin:
+    pk: int
     value: int
 
+
+@dataclass
+class LedgerState:
+    """
+    A snapshot of the ledger state up to some height
+    """
+
+    height: int = None
+    nonce: bytes = None
+    total_stake: int = None
+
+
+@dataclass
+class EpochState:
+    # for details of snapshot schedule please see:
+    # https://github.com/IntersectMBO/ouroboros-consensus/blob/fe245ac1d8dbfb563ede2fdb6585055e12ce9738/docs/website/contents/for-developers/Glossary.md#epoch-structure
+
+    # The stake distribution snapshot is taken at the beginning of the previous epoch
+    stake_distribution_snapshot: LedgerState
+
+    # The nonce snapshot is taken 7k/f slots into the previous epoch
+    nonce_snapshot: LedgerState
+
+    def stake_distribution(self) -> int:
```

Comment:
nit: `stake_distribution` is not a good name for something returning total stake

Reply:
Agreed, I borrowed the name from ouroborous, do you think it's worth changing to `total_stake` or shall we keep in line with their terminology?

Reply:
Do they actually call stake distribution the total stake?

Reply:
https://github.com/IntersectMBO/ouroboros-consensus/blob/fe245ac1d8dbfb563ede2fdb6585055e12ce9738/docs/website/contents/for-developers/Glossary.md#stake-and-delegation

Reading this again, no I don't think that stake distribution == total stake, I'll rename this function 👍